File: src/bytecode/structs/Opcode.ts
import { InstructionType } from "../enums/InstructionType";
import { OpcodeType } from "../enums/OpcodeType";
import { Instruction } from "./Instruction";
import { OpAdd } from "./opcodes/OpAdd";
import { OpCall } from "./opcodes/OpCall";
import { OpClose } from "./opcodes/OpClose";
import { OpClosure } from "./opcodes/OpClosure";
import { OpConcat } from "./opcodes/OpConcat";
import { OpDiv } from "./opcodes/OpDiv";
import { OpEq } from "./opcodes/OpEq";
import { OpForLoop } from "./opcodes/OpForLoop";
import { OpForPrep } from "./opcodes/OpForPrep";
import { OpGetGlobal } from "./opcodes/OpGetGlobal";
import { OpGetTable } from "./opcodes/OpGetTable";
import { OpGetUpval } from "./opcodes/OpGetUpval";
import { OpJmp } from "./opcodes/OpJmp";
import { OpLe } from "./opcodes/OpLe";
import { OpLen } from "./opcodes/OpLen";
import { OpLoadBool } from "./opcodes/OpLoadBool";
import { OpLoadConst } from "./opcodes/OpLoadConst";
import { OpLoadNil } from "./opcodes/OpLoadNil";
import { OpLt } from "./opcodes/OpLt";
import { OpMod } from "./opcodes/OpMod";
import { OpMove } from "./opcodes/OpMove";
import { OpMul } from "./opcodes/OpMul";
import { OpNewTable } from "./opcodes/OpNewTable";
import { OpNot } from "./opcodes/OpNot";
import { OpPow } from "./opcodes/OpPow";
import { OpReturn } from "./opcodes/OpReturn";
import { OpSelf } from "./opcodes/OpSelf";
import { OpSetGlobal } from "./opcodes/OpSetGlobal";
import { OpSetList } from "./opcodes/OpSetList";
import { OpSetTable } from "./opcodes/OpSetTable";
import { OpSetUpval } from "./opcodes/OpSetUpval";
import { OpSub } from "./opcodes/OpSub";
import { OpTailCall } from "./opcodes/OpTailCall";
import { OpTest } from "./opcodes/OpTest";
import { OpTestSet } from "./opcodes/OpTestSet";
import { OpTForLoop } from "./opcodes/OpTForLoop";
import { OpUnm } from "./opcodes/OpUnm";
import { OpVarArg } from "./opcodes/OpVarArg";

const opcodeMappings: Record<OpcodeType, any> = {
  [OpcodeType.Move]: OpMove,
  [OpcodeType.LoadConst]: OpLoadConst,
  [OpcodeType.LoadBool]: OpLoadBool,
  [OpcodeType.LoadNil]: OpLoadNil,
  [OpcodeType.GetUpval]: OpGetUpval,
  [OpcodeType.GetGlobal]: OpGetGlobal,
  [OpcodeType.GetTable]: OpGetTable,
  [OpcodeType.SetGlobal]: OpSetGlobal,
  [OpcodeType.SetUpval]: OpSetUpval,
  [OpcodeType.SetTable]: OpSetTable,
  [OpcodeType.NewTable]: OpNewTable,
  [OpcodeType.Self]: OpSelf,
  [OpcodeType.Add]: OpAdd,
  [OpcodeType.Sub]: OpSub,
  [OpcodeType.Mul]: OpMul,
  [OpcodeType.Div]: OpDiv,
  [OpcodeType.Mod]: OpMod,
  [OpcodeType.Pow]: OpPow,
  [OpcodeType.Unm]: OpUnm,
  [OpcodeType.Not]: OpNot,
  [OpcodeType.Len]: OpLen,
  [OpcodeType.Concat]: OpConcat,
  [OpcodeType.Jmp]: OpJmp,
  [OpcodeType.Eq]: OpEq,
  [OpcodeType.Lt]: OpLt,
  [OpcodeType.Le]: OpLe,
  [OpcodeType.Test]: OpTest,
  [OpcodeType.TestSet]: OpTestSet,
  [OpcodeType.Call]: OpCall,
  [OpcodeType.TailCall]: OpTailCall,
  [OpcodeType.Return]: OpReturn,
  [OpcodeType.ForLoop]: OpForLoop,
  [OpcodeType.ForPrep]: OpForPrep,
  [OpcodeType.TForLoop]: OpTForLoop,
  [OpcodeType.SetList]: OpSetList,
  [OpcodeType.Close]: OpClose,
  [OpcodeType.Closure]: OpClosure,
  [OpcodeType.VarArg]: OpVarArg,
};

export abstract class Opcode {
  abstract instruction: Instruction;

  abstract getObfuscated(): string;
}

File: Obfuscator/Bytecode/Structs/Opcode.cs
using System.Reflection;

abstract class Opcode
{
    private static readonly InstructionType[] instructionMappings = {
        InstructionType.ABC,
        InstructionType.ABx,
        InstructionType.ABC,
        InstructionType.ABC,
        InstructionType.ABC,
        InstructionType.ABx,
        InstructionType.ABC,
        InstructionType.ABx,
        InstructionType.ABC,
        InstructionType.ABC,
        InstructionType.ABC,
        InstructionType.ABC,
        InstructionType.ABC,
        InstructionType.ABC,
        InstructionType.ABC,
        InstructionType.ABC,
        InstructionType.ABC,
        InstructionType.ABC,
        InstructionType.ABC,
        InstructionType.ABC,
        InstructionType.ABC,
        InstructionType.ABC,
        InstructionType.AsBx,
        InstructionType.ABC,
        InstructionType.ABC,
        InstructionType.ABC,
        InstructionType.ABC,
        InstructionType.ABC,
        InstructionType.ABC,
        InstructionType.ABC,
        InstructionType.ABC,
        InstructionType.AsBx,
        InstructionType.AsBx,
        InstructionType.ABC,
        InstructionType.ABC,
        InstructionType.ABC,
        InstructionType.ABx,
        InstructionType.ABC,
   };

    private static readonly Func<int, Opcode>[] opcodeConstructors = {
        (int data) => new OpMove(data),
        (int data) => new OpLoadConst(data),
        (int data) => new OpLoadBool(data),
        (int data) => new OpLoadNil(data),
        (int data) => new OpGetUpval(data),
        (int data) => new OpGetGlobal(data),
        (int data) => new OpGetTable(data),
        (int data) => new OpSetGlobal(data),
        (int data) => new OpSetUpval(data),
        (int data) => new OpSetTable(data),
        (int data) => new OpNewTable(data),
        (int data) => new OpSelf(data),
        (int data) => new OpAdd(data),
        (int data) => new OpSub(data),
        (int data) => new OpMul(data),
        (int data) => new OpDiv(data),
        (int data) => new OpMod(data),
        (int data) => new OpPow(data),
        (int data) => new OpUnm(data),
        (int data) => new OpNot(data),
        (int data) => new OpLen(data),
        (int data) => new OpConcat(data),
        (int data) => new OpJmp(data),
        (int data) => new OpEq(data),
        (int data) => new OpLt(data),
        (int data) => new OpLe(data),
        (int data) => new OpTest(data),
        (int data) => new OpTestSet(data),
        (int data) => new OpCall(data),
        (int data) => new OpTailCall(data),
        (int data) => new OpReturn(data),
        (int data) => new OpForLoop(data),
        (int data) => new OpForPrep(data),
        (int data) => new OpTForLoop(data),
        (int data) => new OpSetList(data),
        (int data) => new OpClose(data),
        (int data) => new OpClosure(data),
        (int data) => new OpVarArg(data),
   };

    public int opcode;
    public InstructionType instructionType;

    public int dataA;
    public int dataB;
    public int dataC;

    public Opcode(int data)
    {
        opcode = data & 0x3f;

        instructionType = instructionMappings[opcode];

        dataA = (data >> 6) & 0xff;

        switch (instructionType)
        {
            case InstructionType.ABC:
                //Why Lua?
                dataB = (data >> 23) & 0x1ff;
                dataC = (data >> 14) & 0x1ff;
                break;
            case InstructionType.ABx:
                dataB = (data >> 14) & 0x3ffff;
                dataC = -1;
                break;
            case InstructionType.AsBx:
                dataB = ((data >> 14) & 0x3ffff) - 131071;
                dataC = -1;
                break;
        }
    }

    public static Opcode Create(int code, int data)
    {
        return opcodeConstructors[code].Invoke(data);
    }

    public virtual string getObfuscatated()
    {
        throw new NotImplementedException();
    }
    public virtual void mutate()
    {
        throw new NotImplementedException();
    }
}